### PR TITLE
feat: Add option to use own NSURLSession for transport

### PR DESCRIPTION
**Pull request by dKasabwala**
…on configuration.
#skip-changelog
## :scroll: Description

- We are a current customer of sentry, and we use custom socks proxy on our end, we have no way to set up socks proxy for sentry services. we have added change to provide hook with URL Session. 

## :bulb: Support Socks proxy and private session configuration with custom SSL Certs.

- Solve Custom SSL pinning 
- Provide support Socks Proxy

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [x] Review from the native team if needed.
- [X] No breaking change or entry added to the changelog.
- [X] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: Sources/Sentry/SentryTransportFactory.m [hunk 1]
```diff
@@ -29,9 +29,18 @@ @implementation SentryTransportFactory
...
+    
+    NSURLSession *session;
+    
+    if (options.urlSession) {
+     
+        session = options.urlSession;
+    } else {
+        session = [NSURLSession sessionWithConfiguration:configuration
+                                                              delegate:options.urlSessionDelegate
+                                                        delegateQueue:nil];
```
Comment (brustolin) [inline Sources/Sentry/SentryTransportFactory.m:41]: m: You could move lines 30, 31 in here.

File: Sources/Sentry/Public/SentryOptions.h [hunk 2]
```diff
@@ -8,6 +8,8 @@ NS_ASSUME_NONNULL_BEGIN
 NS_SWIFT_NAME(Options)
 @interface SentryOptions : NSObject
 
+
```
Comment (brustolin) [inline Sources/Sentry/Public/SentryOptions.h:11]: h: We need description of this new property like the others.

**Review by brustolin — COMMENTED** [review]
This change is pretty simple and gives the user more power, which in my opinion make sense in here.

SentryOptions has a `- (BOOL)validateOptions:(NSDictionary<NSString *, id> *)options
       didFailWithError:(NSError *_Nullable *_Nullable)error` method that need to be updated to accommodate the new property.

We also need some tests to make sure the new property its being used properly.

File: Sources/Sentry/Public/SentryOptions.h (hunk 2 above)
Comment (dKasabwala) [reply Sources/Sentry/Public/SentryOptions.h:14]: Updated description.

File: Sources/Sentry/SentryTransportFactory.m (hunk 1 above)
Comment (dKasabwala) [reply Sources/Sentry/SentryTransportFactory.m:41]: Moved `ephemeralSessionConfiguration` inside else statement.

**Comment by dKasabwala** [comment]
- Added unit tests around new options for urlSession
- updated `- (BOOL)validateOptions:(NSDictionary<NSString *, id> *)options didFailWithError:(NSError *_Nullable *_Nullable)error`
- new property description been added as well.

File: Sources/Sentry/SentryTransportFactory.m [hunk 3]
```diff
@@ -27,11 +27,19 @@ @implementation SentryTransportFactory
...
 {
-    NSURLSessionConfiguration *configuration =
-        [NSURLSessionConfiguration ephemeralSessionConfiguration];
-    NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration
-                                                          delegate:options.urlSessionDelegate
-                                                     delegateQueue:nil];
+
+    NSURLSession *session;
+
+    if (options.urlSession) {
```
Comment (philipphofmann) [inline Sources/Sentry/SentryTransportFactory.m:33]: `m`: Please add some tests to the `SentryTransportFactoryTests` for this new functionality.

Comment (dKasabwala) [reply Sources/Sentry/SentryTransportFactory.m:33]: Updated unit test to cover 100% coverage for this file.

**Comment by dKasabwala** [comment]
@philipphofmann Please get this approved, All checks are complete, we do need these changes next release version.

**Review by philipphofmann — COMMENTED** [review]
The unit tests in CI are failing because the unit test don't compile @dKasabwala. You can ignore the other failing CI jobs. They fail due to our CI setup.

**Comment by dKasabwala** [comment]
Fix up compiler error

**Review by philipphofmann — APPROVED** [review]
LGTM

**Pull request merged**